jenkins/bootstrap.py

```diff
@@ -0,0 +1,514 @@
+    AppendResult(gsutil, paths.result_cache, build, version, success)
+    if paths.pr_result_cache:
+        AppendResult(gsutil, paths.pr_result_cache, build, version, success)
+
+    data = {
+        'timestamp': time.time(),
+        'result': 'SUCCESS' if success else 'FAILURE',
+        'passed': bool(success),
+    }
+    # TODO(rmmh): update tooling to expect metadata in finished.json
+    metadata = os.path.join(paths.artifacts, 'metadata.json')
+    if os.path.isfile(metadata):
+        try:
+            with open(metadata) as fp:
+                val = json.loads(fp.read())
+        except (IOError, ValueError):
+            val = None
+        if val and isinstance(val, dict):
+            data['metadata'] = val
+    gsutil.UploadJson(paths.finished, data)
+
+
+def TestInfra(*paths):
+    """Return path relative to root of test-infra repo."""
+    return os.path.join(ORIG_CWD, os.path.dirname(__file__), '..', *paths)
+
+
+def Node():
+    return os.environ[NODE_ENV]
+
+
+def Version():
+    version_file = 'version'
+    if os.path.isfile(version_file):
+        # e2e tests which download kubernetes use this path:
+        with open(version_file) as fp:
+            return fp.read().strip()
+
+    version_script = 'hack/lib/version.sh'
+    if os.path.isfile(version_script):
+        cmd = [
+            'bash', '-c', (
+"""
+set -o errexit
+set -o nounset
+export KUBE_ROOT=.
+source %s
+kube::version::get_version_vars
+echo $KUBE_GIT_VERSION
+""" % version_script)
+        ]
+        return Subprocess(cmd, output=True).strip()
+
+    return 'unknown'
+
+
+class Paths(object):
+    """Links to remote gcs-paths for uploading results."""
+    def __init__(
+        self,
+        artifacts,  # artifacts folder (in build)
+        build_log,  # build-log.txt (in build)
+        build_path,  # path to build
+        finished,  # finished.json (metadata from end of build)
+        latest,   # latest-build.txt (in job)
+        pr_build_link,  # file containng build_path (in job directory)
+        pr_latest,  # latest-build.txt (in pr job)
+        pr_result_cache,  # jobResultsCache.json (in pr job)
+        result_cache,  # jobResultsCache.json (cache of latest results in job)
+        started,  # started.json  (metadata from start of build)
+    ):
+        self.artifacts = artifacts
+        self.build_log = build_log
+        self.build_path = build_path
+        self.finished = finished
+        self.latest = latest
+        self.pr_build_link = pr_build_link
+        self.pr_latest = pr_latest
+        self.pr_result_cache = pr_result_cache
+        self.result_cache = result_cache
+        self.started = started
+
+
+
+def CIPaths(job, build):
+    base = 'gs://kubernetes-jenkins/logs'
+    latest = os.path.join(base, job, 'latest-build.txt')
+    return Paths(
+        artifacts=os.path.join(base, job, build, 'artifacts'),
+        build_log=os.path.join(base, job, build, 'build-log.txt'),
+        build_path=None,
```

Comment:
minor nit: it's not super obvious to me why this is `None`. (I guess it's because you only use it for the PR link?)